src/compiler/checker.ts

```diff
@@ -4020,7 +4020,10 @@ namespace ts {
                         }
                     });
                 }) : undefined;
-                let res = firstVariableMatch ? [firstVariableMatch, ...additionalContainers, container] : [...additionalContainers, container];
+                let res = firstVariableMatch ? [firstVariableMatch, ...additionalContainers] : [...additionalContainers];
+                if (!container.valueDeclaration || container.valueDeclaration.kind !== SyntaxKind.FunctionDeclaration || !isExpandoFunctionDeclaration(container.valueDeclaration)) {

```

Comment:
I'm not super comfortable with this line of code. My goal was to **preserve** the very same info text that we get when requesting this based on the generated `.d.ts`. We can check this situation [here](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220510#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXzDigxAAUYcAHAZwB4AVAPgAoMAueegSg4G8AoeEPjMe8AG44swANyDhhEMRAA1WFiioM1DqPgBeRhKmz+AXzn8A9FfgALHOJAwsqAObxHz+AANFytRdNbR94YBwQalQAcgx4AHccGABre2cQfn8ScipqZl4zLgA6LNV1YLyuOSA).

As we might notice this **doesn't** include `getProps` as that is being elided in the `.d.ts`.

Maybe I shouldn't try accomplishing this at all, and I should be fine with this "inconsistency", and thus I should just be fine with returning `(property) getProps.createVariants: () => void` here.

IIRC that can be achieved by just handling `SyntaxKind.FunctionDeclaration` [here](https://github.dev/microsoft/TypeScript/blob/9236e39374c0ec9a1e3f9894af4fb9eb34ba0021/src/compiler/checker.ts#L10064). However, I feel that in such a case the function name's probably should be adjusted - if you prefer that solution, let me know and if you could suggest how this should be renamed I would appreciate it (if that's a suitable solution in the first place).

Either way... in an attempt to be consistent about the content of this text between those scenarios I've followed the logic of the `.d.ts` generation and I've found out about the `isExpandoFunctionDeclaration` check being involved. I'm not sure what it exactly means though and including it here changes some existing baselines which isn't good.

At first I've thought that maybe they should actually be adjusted if I'm trying to be "consistent" here but I've analyzed them, checked their `.d.ts` and stuff and I **think** that they shouldn't be changed here (although I'm not completely sure about that, I wasn't exactly sure how properly to validate this for all of those adjusted baselines). So I feel like maybe this check here isn't exactly correct but I'm unsure how to adjust it right now.

Btw. any kind of a condition added to fix this should also cover function expressions case - I probably should add a similar test case for this scenario as it doesn't work either ([here](https://www.typescriptlang.org/play?ts=4.8.0-dev.20220510#code/MYewdgzgLgBMBOBTAhlRAFeIAOEYF4YAeAFQBoA+ACigC4YSBKAimAbwCgY5xoYBzRFEw48hKs3ys2AXy49IsBCjQA1ZPACWyMFDEwJLdnPmDhWXADplqROq069BOElv3tuiAG55SKAFd4MAEhEVwfEwB6SJgACxAAN0QtMH4YROSYAAMbNQ0PPSyYABMQRAgwAHJYAHcQeABrOOTEDlyMCwgqWUZrVzyHTwkfIA)).

Reply:
I have no idea about if the fix is right. The thing I'm more surprised about is the fact that any of this fires in a `.ts` file. IIRC, we're only supposed to support "namespace"-y function declarations at the top level.

Reply:
What kind of support do u mean? I've rechecked all TS versions available in the playground and the pattern seems to be typechecked correctly in all of them (the quick info crashes in all of them though). 🤔 From what I understand this is required to make stuff like `.defaultProps` to work in React.

Reply:
> An expando function property actually equivalent to

Out of curiosity - is "expando function property" specifically about function+namespace combination? Or Special Property Assignment on functions always treated/named as expando too?

> I think the getProps. should be retained; I don't think it matters that much whether it reads function or (property)

My original hesitation before including the namespace was that in the generated `.d.ts` there won't be any namespace here, it's being completely elided:
```ts
// .ts
function createProps<T>(t: T) {
  function getProps() {}
  function createVariants() {}

  getProps.createVariants = createVariants;
  return getProps;
}

// .d.ts
declare function createProps<T>(t: T): {
    (): void;
    createVariants: () => void;
};
```

Reply:
I think that omitting `getProps` is OK here, and that the return type of `createProps` is correct. You can't refer to `getProps` on the return value, so it's fine not to name it via generating a namespace.